### PR TITLE
[com_associations] - use cache for getContentLanguages()

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #15494.

### Summary of Changes
use cache instead of query everytime the available languages

### Testing Instructions
see  #15494

### Expected result
reduce the number of duplicate query

### Actual result
unnecessary duplicate query for get languages

### Additional Comments
reduced this one only
https://github.com/joomla/joomla-cms/issues/15494#issuecomment-343197629

### CODE REVIEW

**Comment by csthomas** [comment]
I would suggest this for `getContentLanguages()`.
```php
public static function getContentLanguages()
 {
	if (empty(static::$languages))
	{
		static::load();
	}

	return static::$languages;
}
```

**Comment by alikon** [comment]
thanks @csthomas   :+1:  more clean now

File: administrator/components/com_associations/helpers/associations.php [hunk 1]
```diff
@@ -452,18 +452,13 @@ private static function getEnabledExtensions()
...
-		else
+		if (empty(static::$languages))
 		{
 			static::load();
-            		$result = static::$languages;
 		}
-
-		return $result;
-	}
+
```
Comment (Quy) [inline administrator/components/com_associations/helpers/associations.php]: Please remove whitespace.

File: administrator/components/com_associations/helpers/associations.php [hunk 2]
```diff
@@ -452,18 +452,13 @@ private static function getEnabledExtensions()
...
 		{
 			static::load();
-            		$result = static::$languages;
 		}
-
-		return $result;
-	}
+ 
+		return static::$languages;
+   	}
```
Comment (Quy) [inline administrator/components/com_associations/helpers/associations.php]: Please fix whitespace.

**Comment by infograf768** [comment]
Queries number has reduced a lot: 186 instead of 235.
as you said, some more could be done. Maybe for another PR?

screen shot 2017-11-10 at 08 23 24

I have seen some cs.

File: administrator/components/com_associations/helpers/associations.php [hunk 3]
```diff
@@ -37,6 +37,14 @@ class AssociationsHelper extends JHelperContent
 	public static $supportedExtensionsList = array();
 
 	/**
+	 * List languages of languages
```
Comment (infograf768) [inline administrator/components/com_associations/helpers/associations.php]: languages of languages ?

File: administrator/components/com_associations/helpers/associations.php [hunk 4]
```diff
@@ -670,4 +672,41 @@ public static function getLanguagefilterPluginId()
...
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected static function load()
+	{
+		$loader = function ()
+		{
+			$db = \JFactory::getDbo();
+			$query = $db->getQuery(true)
+			    ->select($db->quoteName(array('sef', 'lang_code', 'image', 'title', 'published')))
```
Comment (infograf768) [inline administrator/components/com_associations/helpers/associations.php]: 689 and below, replace spaces for indent by tab

File: administrator/components/com_associations/helpers/associations.php [hunk 5]
```diff
@@ -670,4 +672,41 @@ public static function getLanguagefilterPluginId()
...
+		try
+		{
+			static::$languages = $cache->get($loader, array(), __METHOD__);
+		}
+		catch (\JCacheException $e)
+		{
+			static::$languages = $loader();
+		}
+
+    		return true;
```
Comment (infograf768) [inline administrator/components/com_associations/helpers/associations.php]: Indent: delete spaces

**Review by infograf768 — CHANGES_REQUESTED** [review]

**Comment by alikon** [comment]
yes matter for another PR..... 
i'll fix the cs glitch....

File: administrator/components/com_associations/helpers/associations.php (hunk 3 above)
Comment (alikon) [reply administrator/components/com_associations/helpers/associations.php]: changed

File: administrator/components/com_associations/helpers/associations.php (hunk 4 above)
Comment (alikon) [reply administrator/components/com_associations/helpers/associations.php]: replaced

File: administrator/components/com_associations/helpers/associations.php (hunk 5 above)
Comment (alikon) [reply administrator/components/com_associations/helpers/associations.php]: idented

**Comment by alikon** [comment]
hope cs fixed now

**Comment by andrepereiradasilva** [comment]
hi guys, just stoped by to say: why not reuse the already available method in the library 
```php
Joomla\CMS\Language\LanguageHelper::getContentLanguages($checkPublished = true, $checkInstalled = true, $pivot = 'lang_code', $orderField = null, $orderDirection = null)
```

See https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Language/LanguageHelper.php#L336

**Comment by alikon** [comment]
wellcome back @andrepereiradasilva , we are missing you ,

i'll investigate further your suggestion, i hope in the week-end....

**Review by infograf768 — APPROVED** [review]

**Comment by alikon** [comment]
applyed the tips from @andrepereiradasilva 
please re-test

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 6f6c1d329d51a7fb546ca538c05b128eaf5a3e64<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by infograf768** [comment]
I have tested this item :red_circle: unsuccessfully on 6f6c1d329d51a7fb546ca538c05b128eaf5a3e64<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by infograf768** [comment]
Changed my test to unsuccessful.

I get an error
`Call to undefined method AssociationsHelper::getContentLanguages()`
when trying to load the side to side page

**Comment by alikon** [comment]
Silly me, I'll do the needed change later sorry


Il 11 nov 2017 9:45 AM, "infograf768" <notifications@github.com> ha scritto:

**Comment by alikon** [comment]
re-re-test please

File: administrator/components/com_associations/helpers/associations.php [hunk 6]
```diff
@@ -444,19 +445,8 @@ private static function getEnabledExtensions()
...
-			->from($db->quoteName('#__languages'))
-			->where($db->quoteName('published') . ' != -2')
-			->order($db->quoteName('ordering') . ' ASC');
-
-		$db->setQuery($query);
-
-		return $db->loadObjectList('lang_code');
-	}
+		return LanguageHelper::getContentLanguages();
+   	}
```
Comment (infograf768) [inline administrator/components/com_associations/helpers/associations.php]: delete unwanted spaces line 449

Comment (infograf768) [reply administrator/components/com_associations/helpers/associations.php]: otherwise, pr is fine

Comment (alikon) [reply administrator/components/com_associations/helpers/associations.php]: done

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on e935a9d6e1afe2ead3e7962e83cc4273d0b154c5<br /><br />Perfect this time. :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by alikon** [comment]
perfection is my current challenge, unfortunately i already know that i cant win :smile:

**Comment by ghost** [comment]
@alikon can you please give Test Instructions?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by alikon** [comment]
1. start with a multilanguage site with > 5 langauges 
2. add some content in en-GB for example
3. enable debug plugin
4. com_associations  select articles and language en-gb
5. look at database tab of debug plugin and take note of much query are executed and how much are duplicate
6. apply PR
7 redo 4., 5. you should notice less query and less duplicate

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on e935a9d6e1afe2ead3e7962e83cc4273d0b154c5
#### Without PR

#### With PR

Thanks for Instructions @alikon 
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Comment by joomdonation** [comment]
Just want to point out that there is difference between the data returned by old code and new code. With the original code, all content languages (including unpublished) will be returned, with new code, only published content languages will be returned

Also, if we decide to use new code, I think it is better to call the new method directly instead of still calling AssociationsHelper::getContentLanguages (because this method is now basically, an alias of LanguageHelper::getContentLanguages() )

**Comment by infograf768** [comment]
@joomdonation 

You are totally correct. I take off RTC.
We should be able to prepare a new Content Language but not put it online until Published.

With this patch, it is now impossible to do anything for such a content language in com_associations.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by infograf768** [comment]
In fact I guess the solution is easy
`return LanguageHelper::getContentLanguages(false, true);` is enough.

**Comment by alikon** [comment]
applyed the tips from @joomdonation

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 69d811de89f2d7677b5772b3c92cf46b7bb1ae3b<br /><br />Works fine now even if the Content Language is unpublished.

Only drawback is that it also proposes those which are in the Trash.
That case is not taken care of in `LanguageHelper::getContentLanguages()`<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@infograf768 reggarding the trashed language, true. the language helper method could be improved (while preserving B/C).

Something like this. Not tested. Replaced $checkPublished (boolean) argument for $publishedStates (array) and using a B/C layer for that parameter.

```php
/**
 * Get a list of content languages.
 *
 * @param   array    $publishedStates  Array with the content language published states. Null or empty array for all.
 * @param   boolean  $checkInstalled   Check if the content language is installed.
 * @param   string   $pivot            The pivot of the returning array.
 * @param   string   $orderField       Field to order the results.
 * @param   string   $orderDirection   Direction to order the results.
 *
 * @return  array  Array of the content languages.
 *
 * @since   3.7.0
 */
public static function getContentLanguages($publishedStates = array(1), $checkInstalled = true, $pivot = 'lang_code', $orderField = null, $orderDirection = null)
{
	// B/C layer. Before __DEPLOY_VERSION__.
	if ($publishedStates === true)
	{
		$publishedStates = array(1);
	}
	elseif ($publishedStates === false)
	{
		$publishedStates = null;
	}

	// [code removed for brevity...]

	// Check if the language is published, if needed.
	if ($publishedStates !== null && $publishedStates !== array())
	{
		foreach ($languages as $key => $language)
		{
			if (in_array((int) $language->published, $publishedStates, true) === false)
			{
				unset($languages[$key]);
			}
		}
	}

	// [code removed for brevity...]
}

```

then, for published/unpublished, you could call it like this 

```php
LanguageHelper::getContentLanguages(array(0, 1), true);
```
BTW @alikon unless things changed recently i don't think it's B/C to completly remove a public method in 3.x, you should proxy it to languagehelper method and them put a deprecated 4.0 message.

**Comment by alikon** [comment]
managed the trashed lanugages

**Comment by infograf768** [comment]
Works here

**Comment by joomdonation** [comment]
@alikon If I am not mistaken, the changes you made to \Joomla\CMS\Language\LanguageHelper::getContentLanguages() will cause B/C issue. Before the change, if someone calls \Joomla\CMS\Language\LanguageHelper::getContentLanguages() , the result will include trashed content languages. After your changes, it won't include trashed content languages anymore.

I think the propose change from @andrepereiradasilva is better. It extends the method to allow us to get content languages we want based on passed publish states. I made a PR to your branch https://github.com/alikon/joomla-cms/pull/38, please review it and if you agree, merge it so that we can have this issue sorted.

**Comment by alikon** [comment]
thanks folks for teamworking, hope now issues are solved in a B/C way

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on e3d1b462947ea45b5263d80c7645249d33a46f2b<br /><br />TBH, folks, I see absolutely no reason why anyone would list Trashed Content Languages.
In this case, but also everywhere else in J where we use the method.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on f438af01ca1a894d0599ff8bb1060f1e7cf27be2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by ghost** [comment]
@infograf768 can you please retest?

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on f438af01ca1a894d0599ff8bb1060f1e7cf27be2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Comment by infograf768** [comment]
RTC. Thanks to all.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18544">issues.joomla.org/tracker/joomla-cms/18544</a>.</sub>

**Pull request merged**